README.md
# WWW::Bugzilla::BugTree [![Build Status](https://secure.travis-ci.org/plicease/WWW-Bugzilla-BugTree.png)](http://travis-ci.org/plicease/WWW-Bugzilla-BugTree)

Fetch a tree of bugzilla bugs blocking a bug

# SYNOPSIS

    use WWW::Bugzilla::BugTree;
    
    my $tree = WWW::Bugzilla::BugTree->new(
      url => 'http://bugzilla',
    );
    
    # $bug isa WWW::Bugzilla::BugTree::Bug
    my $bug = $tree->fetch(749922);
    print $bug;
    foreach my $subbug (@{ $bug->children })
    {
      print $bug;
    }

# DESCRIPTION

This module provides a way to fetch a tree of dependent bugs from Bugzilla.
You give it a bug id and it returns a tree of all the bugs that bug depends
on (or all the bugs that are blocking your bug).  I wrote this to use the
`XML` output of Bugzilla's `show_bug.cgi` page because we are still using
Bugzilla 3.6, which doesn't provide dependency information via its API, which
would probably be faster.

There is also a companion script [bug\_tree](https://metacpan.org/pod/bug_tree) which will print out the tree
for you with pretty colors indicating each bug's status.

# ATTRIBUTES

## ua

    my $lwp = $tree->ua;

Instance of [LWP::UserAgent](https://metacpan.org/pod/LWP::UserAgent) used to fetch information from the
bugzilla server.

## url

    my$url = $tree->url

The URI of the bugzilla server.  You may pass in to the constructor
either a string or a [URI](https://metacpan.org/pod/URI) object.  If you use a string it will
be converted into a [URI](https://metacpan.org/pod/URI).

If not provided it falls back to using the `BUG_TREE_URL` environment
variable, and if that isn't set it uses this bugzilla provided for
testing:

[Bugzilla v4.2](https://landfill.bugzilla.org/bugzilla-4.2-branch)

# METHODS

## fetch

    my $bug = $tree->fetch($id);

Fetch the bug tree for the bug specified by the given `id`.  Returns
an instance of [WWW::Bugzilla::BugTree::Bug](https://metacpan.org/pod/WWW::Bugzilla::BugTree::Bug).

## clear\_cache

    $tree->clear_cache;

Clears out the cache.

# SEE ALSO

[bug\_tree](https://metacpan.org/pod/bug_tree), [WWW::Bugzilla::BugTree::Bug](https://metacpan.org/pod/WWW::Bugzilla::BugTree::Bug)

# AUTHOR

Graham Ollis <plicease@cpan.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.08      2018-06-25 17:00:11 -0400
  - disable fetch.t as landfill has been completely decomissioned.

0.07      2015-12-17 08:24:12 -0500
  - Do not test against landfill 4.0 anymore, it is gonners!
  - Use Ref::Util for ref testing

0.06      2015-11-22 15:42:45 -0500
  - Pull required Perl back to 0.012

0.05      2015-06-11 13:22:08 -0400
  - No longer test against 3.6 landfill since it is no longer available
  - Require Perl 5.14

0.04      2015-03-03 23:44:58 -0500
  - Require Moo 2.0, as older versions inadvertently turn on
    fatal warnings

0.03      2013-10-10 10:49:40 -0400
  - use proxy if http_proxy is set

0.02      2013-08-07 06:42:05 America/New_York
  - added BUG_TREE_USER environment variable for bug_tree script

0.01      2013-08-07 06:31:21 America/New_York
  - initial version


lib/WWW/Bugzilla/BugTree/Bug.pm
package WWW::Bugzilla::BugTree::Bug;

use strict;
use warnings;
use 5.012;
use Moo;
use XML::Simple qw( XMLin );
use overload '""' => sub { shift->as_string };

# ABSTRACT: A bug tree returned from WWW::Bugzilla::BugTree
# VERSION

=head1 DESCRIPTION

This class represents an individual bug returned from L<WWW::Bugzilla::BugTree>'s C<fetch> method.
It is also a tree since it has a C<children> accessor which returns the list of bugs that block
this bug.

=head1 ATTRIBUTES

=head2 url

 my $url = $bug->url;

The URL of the bug.

=cut

has url => (
  is       => 'ro',
  required => 1,
);

=head2 res

 my $res = $bug->res;

The raw L<HTTP::Response> object for the bug.

=cut

has res => (
  is       => 'ro',
  required => 1,
);

=head2 id

 my $id = $bug->id;

The bug id for the bug.

=cut

has id => (
  is       => 'ro',
  required => 1,
);

has as_hashref => (
  is       => 'ro',
  init_arg => undef,
  lazy     => 1,
  default  => sub {
    no warnings;
    XMLin(shift->res->decoded_content);
  },
);

=head2 children

 my @children = $bug->children->@*;

The list of bugs that are blocking this one.
This is a list of L<WWW::Bugzilla::BugTree::Bug> objects.

=cut

has children => (
  is       => 'ro',
  init_arg => undef,
  default  => sub { [] },
);

=head2 as_string

 my $string = $bug->as_string;
 my $string = "$bug";

Returns a human readable form of the string in the form of

 "id status (resolution) subject"

if it has been resolved, and 

 "id status subject"

otherwise.

=cut

sub as_string
{
  my($self) = @_;
  my $id         = $self->id;
  my $status     = $self->as_hashref->{bug}->{bug_status};
  my $subject    = $self->as_hashref->{bug}->{short_desc};
  my $resolution = $self->as_hashref->{bug}->{resolution};
  undef $resolution if ref $resolution;
  $resolution ? "$id $status ($resolution) $subject" : "$id $status $subject";
}

# undocumented function
sub summary_tree
{
  my($self) = @_;
  
  [ $self->as_string, @{ $self->children } > 0 ? map { $_->summary_tree } @{ $self->children } : () ];
}

1;

=head1 SEE ALSO

L<bug_tree>, L<WWW::Bugzilla::BugTree>

=cut


lib/WWW/Bugzilla/BugTree.pm
package WWW::Bugzilla::BugTree;

use strict;
use warnings;
use Ref::Util qw( is_arrayref );
use 5.012;
use Moo;

# ABSTRACT: Fetch a tree of bugzilla bugs blocking a bug
# VERSION

=head1 SYNOPSIS

 use WWW::Bugzilla::BugTree;
 
 my $tree = WWW::Bugzilla::BugTree->new(
   url => 'http://bugzilla',
 );
 
 # $bug isa WWW::Bugzilla::BugTree::Bug
 my $bug = $tree->fetch(749922);
 print $bug;
 foreach my $subbug (@{ $bug->children })
 {
   print $bug;
 }

=head1 DESCRIPTION

This module provides a way to fetch a tree of dependent bugs from Bugzilla.
You give it a bug id and it returns a tree of all the bugs that bug depends
on (or all the bugs that are blocking your bug).  I wrote this to use the
C<XML> output of Bugzilla's C<show_bug.cgi> page because we are still using
Bugzilla 3.6, which doesn't provide dependency information via its API, which
would probably be faster.

There is also a companion script L<bug_tree> which will print out the tree
for you with pretty colors indicating each bug's status.

=head1 ATTRIBUTES

=head2 ua

 my $lwp = $tree->ua;

Instance of L<LWP::UserAgent> used to fetch information from the
bugzilla server.

=cut

has ua => (
  is      => 'ro',
  lazy    => 1,
  default => sub {
    require LWP::UserAgent;
    my $ua = LWP::UserAgent->new;
    $ua->env_proxy;
    $ua;
  },
);

=head2 url

 my$url = $tree->url

The URI of the bugzilla server.  You may pass in to the constructor
either a string or a L<URI> object.  If you use a string it will
be converted into a L<URI>.

If not provided it falls back to using the C<BUG_TREE_URL> environment
variable, and if that isn't set it uses this bugzilla provided for
testing:

L<Bugzilla v4.2|https://landfill.bugzilla.org/bugzilla-4.2-branch>

=cut

my $default_url = $ENV{BUG_TREE_URL} // "https://landfill.bugzilla.org/bugzilla-4.2-branch";

has url => (
  is      => 'ro',
  lazy    => 1,
  default => sub {
    require URI;
    URI->new($default_url);
  },
  coerce  => sub {
    ref $_[0] ? $_[0] : do { require URI; URI->new($_[0] // $default_url) },
  },
);

has _cache => (
  is       => 'ro',
  default  => sub { { } },
  init_arg => undef,  
);

=head1 METHODS

=head2 fetch

 my $bug = $tree->fetch($id);

Fetch the bug tree for the bug specified by the given C<id>.  Returns
an instance of L<WWW::Bugzilla::BugTree::Bug>.

=cut

sub fetch
{
  my($self, $bug_id) = @_;
  
  return $self->_cache->{$bug_id}
    if exists $self->_cache->{$bug_id};
  
  my $url = $self->url->clone;
  $url->path((do { my $path = $url->path; $path =~ s{/$}{}; $path }) . "/show_bug.cgi");
  $url->query_form(
    id    => $bug_id,
    ctype => 'xml',
  );
  
  my $res = $self->ua->get($url);  
  
  die $url . " " . $res->status_line
    unless $res->is_success;

  require WWW::Bugzilla::BugTree::Bug;  
  my $b = WWW::Bugzilla::BugTree::Bug->new(
    url => $url,
    res => $res,
    id  => $bug_id,
  );
  
  $self->_cache->{$bug_id} = $b;
  
  my $dependson = $b->as_hashref->{bug}->{dependson};
  $dependson = [] unless defined $dependson;
  $dependson = [ $dependson ]
    unless is_arrayref $dependson;
    
  @{ $b->children } = map { $self->fetch($_) } sort @$dependson;
  
  $b;
}

=head2 clear_cache

 $tree->clear_cache;

Clears out the cache.

=cut

sub clear_cache
{
  my($self) = @_;
  %{ $self->_cache } = ();
}

1;

=head1 SEE ALSO

L<bug_tree>, L<WWW::Bugzilla::BugTree::Bug>

=cut

